consensus/src/transaction_shuffler/use_case_aware/delayed_queue.rs

```diff
@@ -0,0 +1,503 @@
+    input_idx: InputIdx,
+    account_by_delay: BTreeMap<DelayKey, AccountAddress>,
+}
+
+impl UseCase {
+    fn new_empty(try_delay_till: OutputIdx, input_idx: InputIdx) -> Self {
+        Self {
+            try_delay_till,
+            input_idx,
+            account_by_delay: BTreeMap::new(),
+        }
+    }
+
+    fn new_with_account<Txn>(
+        try_delay_till: OutputIdx,
+        address: AccountAddress,
+        account: &Account<Txn>,
+    ) -> Self
+    where
+        Txn: UseCaseAwareTransaction,
+    {
+        let mut account_by_delay = BTreeMap::new();
+        account_by_delay.strict_insert(account.delay_key(), address);
+        Self {
+            try_delay_till,
+            input_idx: account.input_idx,
+            account_by_delay,
+        }
+    }
+
+    fn is_empty(&self) -> bool {
+        self.account_by_delay.is_empty()
+    }
+
+    fn delay_key(&self) -> DelayKey {
+        // If head account will be ready later than the use case itself, respect that.
+        let try_delay_till = std::cmp::max(
+            self.try_delay_till,
+            self.account_by_delay
+                .first_key_value()
+                .map_or(0, |(k, _)| k.try_delay_till),
+        );
+
+        DelayKey {
+            try_delay_till,
+            input_idx: self.input_idx,
+        }
+    }
+
+    fn expect_pop_head_account(&mut self) -> (DelayKey, AccountAddress) {
+        let (account_delay_key, address) = self.account_by_delay.pop_first().expect("Must exist.");
+        if let Some((next_account_delay_key, _)) = self.account_by_delay.first_key_value() {
+            self.input_idx = next_account_delay_key.input_idx;
+        }
+        (account_delay_key, address)
+    }
+
+    fn update_try_delay_till(&mut self, try_delay_till: OutputIdx) {
+        self.try_delay_till = try_delay_till;
+    }
+
+    fn add_account<Txn>(&mut self, address: AccountAddress, account: &Account<Txn>)
+    where
+        Txn: UseCaseAwareTransaction,
+    {
+        let account_delay_key = account.delay_key();
+        self.account_by_delay
+            .strict_insert(account_delay_key, address);
+        let (_, head_address) = self.account_by_delay.first_key_value().unwrap();
+        if head_address == &address {
+            self.input_idx = account_delay_key.input_idx;
+        }
+    }
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct DelayedQueue<Txn> {
+    accounts: HashMap<AccountAddress, Account<Txn>>,
+    use_cases: HashMap<UseCaseKey, UseCase>,
+
+    /// empty accounts (those w/o known delayed txns), kept to track the delay
+    account_placeholders_by_delay: BTreeMap<DelayKey, AccountAddress>,
+    /// empty use cases (those w/o known delayed txns), kept to track the delay
+    use_case_placeholders_by_delay: BTreeMap<DelayKey, UseCaseKey>,
+
+    /// main delay queue, all use cases are non-empty of non-empty accounts
+    use_cases_by_delay: BTreeMap<DelayKey, UseCaseKey>,
```

Comment:
can you explain how does it work with the asymmetry here?  everything else has equivalents for both account and use_case spreading, but this one - the main queue - is only be use_cases? i.e. why do we not need accounts_by_delay as well?

Reply:
The queue is nested, containning all pending txns.
It's a queue of use cases, which are queues of accounts, which are queues of txns. whenever a txn is added or removed from this nested structure, the account changes and the account moves within the same use case or across use cases, and then the use case(s) changes position in the top level queue.

This way we can know immediately which txn is the head (will be ready the earliest). If we have head txns for both the account queue and use case queue we won't even be sure that the overall "head" is one of them.

Reply:
ah I see, so 
```
    accounts: HashMap<AccountAddress, Account<Txn>>,
    use_cases: HashMap<UseCaseKey, UseCase>,
```
are the actual queues as well.

One thought - would it be simpler to inline use_cases inside use_cases_by_delay, i.e. have `use_cases_by_delay` be: `BTreeMap<DelayKey, UseCase>`?


For the current code, I suggest reordering the fields of this struct then, with some more comments:

```
/// Queue of all delayed transactions - transactions already iterated on that are waiting to be placed in the final order.
/// Queue is nested, first `use_cases_by_delay` prioritizes all use cases based on DelayKey. 
/// Each UseCase, in `use_cases`, has a queue (`account_by_delay`) where it prioritizes accounts by DelayKey. 
/// And then each 
#[derive(Debug, Default)]
pub(crate) struct DelayedQueue<Txn> {
    /// main delay queue, all delayed usecases prioritized.
    /// all use cases are non-empty of non-empty accounts
    use_cases_by_delay: BTreeMap<DelayKey, UseCaseKey>,
    /// Details for each use-case, including a `account_by_delay` priority queue
    use_cases: HashMap<UseCaseKey, UseCase>,
    /// List of transactions for each account, prioritized by their initial order
    accounts: HashMap<AccountAddress, Account<Txn>>,

    /// Next two keep delay information for currently empty accounts/usecases

    /// empty accounts (those w/o known delayed txns), kept to track the delay
    account_placeholders_by_delay: BTreeMap<DelayKey, AccountAddress>,
    /// empty use cases (those w/o known delayed txns), kept to track the delay
    use_case_placeholders_by_delay: BTreeMap<DelayKey, UseCaseKey>,

    /// externally set output index; when an item has try_delay_till <= output_idx, it's deemed ready
    output_idx: OutputIdx,

    config: Config,


```

Reply:
> One thought - would it be simpler to inline use_cases inside use_cases_by_delay, i.e. have use_cases_by_delay be: BTreeMap<DelayKey, UseCase>?

Maybe not -- I need to look up the delays by account keys and use case keys, also to see if they are empty. These central places are easier in that case, otherwise I'd need mappings from the keys to the positions in the queues (and distinguish between empty accounts/use cases and non empty ones)